fix(gatsby): Chunk nodes when serializing redux to prevent OOM

pvdz opened the pull request:
We are using `v8.serialize` to write and read the redux state. This is faster than `JSON.parse`. Unfortunately, as reported in #17233, this can lead to a fatal when the contents of the redux state is too big to be serialized to a Buffer (hard max of 2GB). Alternatively, we also hit this problem on large site like a million small md pages.

The solution is to shard the `nodes` property, which holds all the page data. In this change I've added a simple heuristic to determine the max chunk size (mind you, currently that's basically `Infinity`). It will serialize about 11 individual nodes, measure their size, and based on the biggest node determine how many nodes would fit in 1.5GB.

The serialization process is updated to no longer put the `nodes` in the main redux file, but rather sharded over a few specific files. When reading the state from cache, these files are all read and their contents are put together in a single Map again. If there were no nodes files this part does nothing so it's even backwards compatible.

Also fixes the TS type for a page node in redux and exports it so we can reference it explicitly.

blainekasten commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,77 @@
-const file = (): string => `${process.cwd()}/.cache/redux.state`
+const CWD = process.cwd()
+const file = (): string => CWD + `/.cache/redux.state`
+const chunkFilePrefix = (): string => CWD + `/.cache/redux.node.state_`
```
sort of nit, do these need to be functions? They don't compute anything

pvdz replied:
I kind of went with was there tbh :) I'm not sure whether there was a particular reason for them to be functions. I'll make them constants because why not.

blainekasten commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,77 @@
 
-const file = (): string => `${process.cwd()}/.cache/redux.state`
+const CWD = process.cwd()
+const file = (): string => CWD + `/.cache/redux.state`
```
The name of this variable isn't helpful for readability. It specifically caught me off guard on L17 when i saw it's usage. Maybe call it reduxStateFile

pvdz replied:
Same as before; I went with the pattern that was there. I'll change them to constants.

blainekasten commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,77 @@
+    v8.deserialize(readFileSync(file))
+  )
+
+  const nodes: [string, IReduxNode][] = [].concat(...chunks)
```
i think i'm confusing myself. But isn't this just creating a copy of the chunks array? If so, it doesn't look like it's really being used in a way that requires a copy. I'm probably missing something here.

pvdz replied:
Yeah, this is shallow flatten. (Intentially one level) This is not just shallow copying an array :)

This was the best performant way I could think of to concat these potentially huge arrays. (A 1M page site that chunks per 10k nodes will have 100 elements here with each 10k elements).

In theory we could pop one element and use that as the left array of the concat, but I don't think that's a real improvement on any level.

blainekasten commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,77 @@
+  // Max size of a Buffer is 2gb (yeah, we're assuming 64bit system)
+  // https://stackoverflow.com/questions/8974375/whats-the-maximum-size-of-a-node-js-buffer
+  // Use 1.5gb as the target ceiling, allowing for some margin of error
+  return Math.floor((150 * 1024 * 1024 * 1024) / maxSize)
```
This is really clever. I'm assuming a limitation of this system is that if by chance, we grabbed 11 small nodes then the OOM could still be hit? 

Is there a reason for the const `11`?

pvdz replied:
Yes, there is a chance. There is a trade-of between how many nodes we trial and the time spent on it.

I picked 11 as an arbitrary number that is the first prime after 10. Open to suggestions if you can think of a better heuristic :)

pieh commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,77 @@
+  // Pick a chunk size ceiling based on the worst case.
+  // This attempts to prevent small sites with very large pages from OOMing
+  let maxSize = 0
+  for (let i = 0; i < valueCount; i += Math.floor(valueCount / 11)) {
```
I agree with the need for heuristic for that as this is pretty expensive, I just wonder about distribution of nodes / node sizes and how well it will work.

I think this is fine for now (better than not even trying), but potential future changes I could see is trying to pick example nodes of every type maybe in the future (?)

pvdz replied:
Yeah I was hoping that by picking nodes randomly throughout the list it would hit random types of nodes, but there's no guarantee of anything here. You might be right that picking nodes explicitly from each type might improve the heuristic.

pieh replied:
I mean - this is not a blocker - this is better than not having anything at all. That's potential option for the future (if we will need it)

blainekasten commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,77 @@
+        v8.serialize(values.slice(i * chunkSize, i * chunkSize + chunkSize))
+      )
+    }
+  }
```
this is cool 🎉

pieh replied:
With state split across many files we "lose" transactional (implicit) nature of using single file - we potentially can save subset of state now (I think we have try/catch somewhere higher in call stack that would make failure to save non-fatal potentially).

It would be good (IMO) to save files to tmp location and if all of them were successful move things around to final one?

pvdz replied:
Oh yes, I forgot about the transactional state. We (or somebody else?) had a discussion about that. Yeah I'll use a tmp location instead, rename the current folder away, and the tmp folder in, and then destroy the old folder. That should be transactional enough... For now anyways. Thanks for reminding me!

pvdz replied:
Swapping in/out a folder means backwards compat breaking.

To mitigate this I'll support the old file and a new folder. We can drop support for the legacy file when we bump a major. Just don't want any unexpected problems related to switching patch, or even minors.

pieh replied:
I think we might blow away entire cache when we upgrade gatsby version (not sure if we catch gatsby core version changes or just plugin version changes when we chose to invalidate cache). So we might not need to do this (needs verification)

pieh replied:
https://github.com/gatsbyjs/gatsby/blob/27aeb08f5d9bc055245b07247838308e1e404e26/packages/gatsby/src/bootstrap/index.js#L214-L231

Ok, we blow away cache if package.json changed - so this doesn't necesairly will catch cases of updating gatsby core (i.e. just blowing away lock files and reinstalling deps)

pvdz replied:
Arguably we should. We could get the current running gatsby version and bake it into the cache and always invalidate it on a miss-match.

pieh replied:
/agree! (but probably in separate pull request)

pvdz replied:
Probably ;)

Let's wait for green and, if you're alright with the latest changes, merge

pieh commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,77 @@
+  // Pick a chunk size ceiling based on the worst case.
+  // This attempts to prevent small sites with very large pages from OOMing
+  let maxSize = 0
+  for (let i = 0; i < valueCount; i += Math.floor(valueCount / 11)) {
```
I agree with the need for heuristic for that as this is pretty expensive, I just wonder about distribution of nodes / node sizes and how well it will work.

I think this is fine for now (better than not even trying), but potential future changes I could see is trying to pick example nodes of every type maybe in the future (?)

pvdz commented:
Added the mechanism for transactional state. This required a bit of refactoring, please have a look.

In general it works as explained before: a tmp folder is created, the new cache is written to that folder, the old cache folder is renamed to a backup location, the tmp folder is moved into the cache folder position, and then an attempt is made to drop the old folder.

Legacy files are properly read and deleted when a new cache is created. Upgrading should work transparently. Downgrading will not work, because an older version will not know to find the new redux cache folder, so it will just miss the cache. I think that's fine, both for us and our users.

Updated some tests to take new things into account. Feels like I've been mocking half the fs-extra lib now :p

pvdz commented:
(Shuffled the functions around to make TS lint happy. It does not seem to like function calls in sibling functions on the same level unless the called function is declared before it ... weird in the JS world)

pvdz commented:
Ignore the next few pushes. I'm adding some logging to debug the windows run through circleci :clown_face:

pieh commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,155 @@
+}
+
+function readFromLegacyCache(): ICachedReduxState {
+  // TODO: remove this bit with next major bump (fairly benign change but still)
```
```suggestion
  // TODO: remove this bit in v3 (fairly benign change but still)
```
we don't have codified rules to be able to grep things we will want to change (neither keep track of it in some issue), so let's use `v3` as keyword for notes like that (we already use it that in other places). This might not be completely accurate (i.e. we might keep some of behaviour marked with "remove in v3", but at least we can search for them easier)

pvdz replied:
Sure, good point

pieh commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,155 @@
+  try {
+    if (existsSync(legacyCacheFile)) {
+      removeSync(legacyCacheFile)
+    }
```
Worth trying to avoid `existSync` here if we already successfully removed this before? Or it pretty much doesn't matter given rest that is happening here and not worth extra (minor) complexity?

pvdz replied:
Technically the first check it doesn't check whether this file exists but rather whether the redux folder exists and then tries to find the legacy cache file.

> Or it pretty much doesn't matter given rest that is happening here and not worth extra (minor) complexity?

Yeah

pieh commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,155 @@
 
-const file = (): string => `${process.cwd()}/.cache/redux.state`
+const legacyCacheFile = path.join(process.cwd(), `.cache/redux.state`)
+const reduxCacheFolder = path.join(process.cwd(), `.cache/redux`)
```
I remembered why this was a function instead of straight assignment - there was this PR - https://github.com/gatsbyjs/gatsby/pull/19800 I did make a note that it's not guaranteed to continue working in the future, but if there is low effort way to keep theirs usage - that would be awesome

pvdz replied:
I see. I'll add a comment for this as well. Seeing how it's now a functional function (it takes an arg) I think we'll be keeping it a little longer anyways :)

pvdz commented:
Ok. All tests are passing. Ready to go

pieh commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,166 @@
+
+  if (chunks.length) {
+    obj.nodes = new Map(nodes)
+  }
```
Core creates some nodes on its own (things like `Site` node type etc - https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/internal-plugins/internal-data-bridge/gatsby-node.js#L98-L107), so there always should be at least one node cached (if we have any cache that is). 

Would it be good idea to have `else` path here in case something went wrong and there are no nodes read from `.cache/redux` (in this code path `.cache/redux` dir exists) and throw? This should make it more proofed and add some very crude "state consistency check"

_This might be wrong place to handle this concern (as it relies about knowledge that core does indeed always create at least one node)_

pvdz commented:
Ok I've added the `else` case. I've not added more tests because it feels like that's more about the mocks than actual fs operations at that point.

pvdz commented:
The return type is incorrect but after quickly trying to fix that and propagate the type just led into a deep rabbit hole so I'm doing the same as the parent calling code, for now.

pvdz commented:
Ah, that `else` now breaks the tests because they omit the `nodes` property due to loki ... ok ok.

pieh commented:
> Ah, that `else` now breaks the tests because they omit the `nodes` property due to loki ... ok ok.

Ehh, didn't think about loki ... but also great that there were tests that caught it (uff).

I don't know if it's worth to try to keep that code path then. We could make it not to show warning / return empty state for loki, but it's get convoluted

pieh commented:
Different and maybe more robust (still not super robust) test could be to just store number of nodes in `rest` file and compare it to number of nodes that were read from chunked nodes files

pvdz commented:
The breakage is artificial; it's because the test removes the `nodes` property before testing it.

I can just add a loki if-branch for the test and have the redux branch not remove those props. I think that'd be fine?

pvdz commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,175 @@
+
+  const nodesToTest = 11 // Very arbitrary number
+  const valueCount = values.length
+  const step = Math.max(1, Math.ceil(valueCount / nodesToTest))
```
This caused an infinite loop, which is why I'm using `max(1, ...)` now. That way the loop always increments. And since it stops as soon as the index exceeds the item count, that should suffice.

wardpeet commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,175 @@
 
-export const writeToCache = (contents: ICachedReduxState): void => {
-  writeFileSync(file(), v8.serialize(contents))
+function reduxRestFile(dir: string): string {
```
I find restFile such an ugly term XD What about making it more sexy with "global" :joy:

```suggestion
function reduxGlobalFile(dir: string): string {
```

pieh replied:
Also consider "shared" (as in the file is shared for multiple redux substates), maybe?

wardpeet commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,175 @@
+function reduxRestFile(dir: string): string {
+  return path.join(dir, `redux.rest.state`)
+}
+function reduxChunkFilePrefix(dir: string): string {
```
This makes it explicitly for "nodes"
```suggestion
function reduxChunkedNodesFilePrefix(dir: string): string {
```

pvdz replied:
Sure :man_shrugging:  :)

wardpeet commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,175 @@
+
+  if (map) {
+    // Now store the nodes separately, chunk size determined by a heuristic
+    const values: [string, IReduxNode][] = [...map.entries()]
```
I wonder what the memory overhead is of creating an array out of it. The cost should only be the strings because the objects aren't cloned. Is there any value on keeping the Map? (I know we have to build the "slice" function ourselves so might not be worth it)

pvdz replied:
The map will make any lookup by id faster at runtime.

If you can think of a better way to do the slicing here I'm all ears.

One alternative would be to `map.forEach` and build up the buckets manually, but I'm convinced that would be slower due to the manual one-by-one pushes that you'd be doing a million times.

wardpeet replied:
That would be my suggestion 🤣. You're  right, let's not do this

wardpeet commented on packages/gatsby/src/redux/persist.ts:
```diff
@@ -1,12 +1,175 @@
+
+function safelyRenameToBak(reduxCacheFolder: string): string {
+  // Basically try to work around the potential of previous renamed caches
+  // not being removed for whatever reason. _That_ should not be a blocker.
```
why not try to remove the bak file? Why would we want to keep other bak files around if they didn't get deleted? (permissions issues would be weird because we're creating them)

pieh replied:
We do try to remove it later on (check `// Now try to yolorimraf the old cache folder` section). If something wrong happens during writing out new cache, we could potentially try to restore previous one (instead of blowing entire cache away). Tho I don't think we try to restore `bak` cache now in this case (and also don't know if it's safe to do - it should be).

It's not only permission issues here - `fs` can be finicky (I found multiple `fs` problems, but mostly on windows and they were intermittent - just restarting something was "solving" the problem)

wardpeet replied:
Isn't that why we use fs-extra to fix windows retries?

pvdz replied:
As mentioned above, the only reason is to try and do the least amount of dangerous work until the new cache is persisted.

I'm not entirely convinced this makes much of a difference because if a fatal happens the cache is in a questionable state, but I think it helps?

The drawback is temporarily requiring double the space. That much is true.
